core/node/vm_runner/src/process.rs

```diff
@@ -0,0 +1,177 @@
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Context;
+use multivm::interface::L2BlockEnv;
+use tokio::{sync::watch, task::JoinHandle};
+use zksync_dal::{ConnectionPool, Core};
+use zksync_state::ReadStorageFactory;
+use zksync_state_keeper::{
+    BatchExecutor, ExecutionMetricsForCriteria, L2BlockParams, TxExecutionResult, UpdatesManager,
+};
+
+use crate::{storage::StorageLoader, OutputHandlerFactory, VmRunnerIo};
+
+/// VM runner represents a logic layer of L1 batch / L2 block processing flow akin to that of state
+/// keeper. The difference is that VM runner is designed to be run on batches/blocks that have
+/// already been processed by state keeper but still require some extra handling as regulated by
+/// [`OutputHandlerFactory`].
+///
+/// It's responsible for taking unprocessed data from the [`VmRunnerIo`], feeding it into
+/// [`BatchExecutor`] and calling [`OutputHandlerFactory`] on the result of the execution (batch
+/// execution state in the [`UpdatesManager`]).
+///
+/// You can think of VM runner as a concurrent processor of a continuous stream of newly committed
+/// batches/blocks.
+#[derive(Debug)]
+pub struct VmRunner {
+    pool: ConnectionPool<Core>,
+    io: Box<dyn VmRunnerIo>,
+    storage: Arc<dyn ReadStorageFactory>,
+    loader: Arc<dyn StorageLoader>,
+    output_handler_factory: Box<dyn OutputHandlerFactory>,
+    batch_processor: Box<dyn BatchExecutor>,
+}
+
+impl VmRunner {
+    /// Initializes VM runner with its constituents. In order to make VM runner concurrent each
+    /// parameter here needs to support concurrent execution mode. See
+    /// [`ConcurrentOutputHandlerFactory`], [`VmRunnerStorage`].
+    ///
+    /// Caller is expected to provide a component-specific implementation of [`VmRunnerIo`] and
+    /// an underlying implementation of [`OutputHandlerFactory`].
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        io: Box<dyn VmRunnerIo>,
+        // Normally `storage` and `loader` would be the same object, but it is awkward to deal with
+        // them as such due to Rust's limitations on upcasting coercion. See
+        // https://github.com/rust-lang/rust/issues/65991.
```

Comment:
If I'm not missing anything, this can be circumvented by adding a specific upcasting method to `StorageLoader`:

```rust
#[async_trait]
pub trait StorageLoader: ReadStorageFactory {
    fn upcast(self: Arc<Self>) -> Arc<dyn ReadStorageFactory>;
    // ...
}
```

The implementation is trivial:

```rust
#[async_trait]
impl<Io: VmRunnerIo> StorageLoader for VmRunnerStorage<Io> {
    fn upcast(self: Arc<Self>) -> Arc<dyn ReadStorageFactory> {
        self
    }
    // ...
}
```

Reply:
Nice trick, worked like a charm!